modules/openy_gc_shared_content_server/src/SourceMigrationDeriver.php

```diff
@@ -0,0 +1,203 @@
+<?php
+
+namespace Drupal\openy_gc_shared_content_server;
+
+use Drupal\Component\Plugin\Derivative\DeriverBase;
+use Drupal\Component\Plugin\Derivative\DeriverInterface;
+use Drupal\Core\Entity\Query\QueryFactory;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Drupal\openy_gc_shared_content_server\Entity\SharedContentSource;
+use Drupal\Core\Plugin\Discovery\ContainerDeriverInterface;
+
+/**
+ * Class SourceMigrationDeriver.
+ *
+ * @package Drupal\openy_gc_shared_content
+ */
+class SourceMigrationDeriver extends DeriverBase implements DeriverInterface, ContainerDeriverInterface {
+
+  /**
+   * EntityQuery service instance.
+   *
+   * @var \Drupal\Core\Entity\Query\QueryFactory
+   */
+  protected $entityQuery;
+
+  /**
+   * SourceMigrationDeriver constructor.
+   *
+   * @param \Drupal\Core\Entity\Query\QueryFactory $entityQuery
+   *   EntityQuery instance.
+   */
+  public function __construct(QueryFactory $entityQuery) {
+    $this->entityQuery = $entityQuery;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, $base_plugin_id) {
+    return new static(
+      $container->get('entity.query')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function getDerivativeDefinitions($base_plugin_definition) {
+
+    $ids = $this->entityQuery->get('shared_content_source')->execute();
+
+    if (empty($ids)) {
+      return [];
+    }
+
+    $sources = SharedContentSource::loadMultiple($ids);
+    $urls = [];
+    foreach ($sources as $source) {
+      $urls[] = $source->getUrl();
+    }
+
+    $params = [
+      'include' => implode(',', $this->getRemoteRelationshipsList($base_plugin_definition)),
+      'sort[sortByDate][path]' => 'created',
+      'sort[sortByDate][direction]' => 'DESC',
+      'filter[status]' => 1,
+      // @TODO add shared_content_filter once it will be at test servers.
+    ];
+
+    $jsonapi_uri = '/jsonapi/node/' . $base_plugin_definition['source']['entity_type'] . '?' . http_build_query($params);
+
+    foreach ($urls as $url) {
+      $url_long = $url . $jsonapi_uri;
+      $derivative = $this->getDerivativeValues($base_plugin_definition, $url_long, $url);
+      $this->derivatives[$this->getKey($url)] = $derivative;
+    }
+
+    return $this->derivatives;
+  }
+
+  /**
+   * Private method that dynamically updates migration source.
+   *
+   * @param array $base_plugin_definition
+   *   Plugin settings.
+   * @param string $url_long
+   *   Url with request part.
+   * @param string $url
+   *   Dynamic url for every Virtual Y content source.
+   *
+   * @return array
+   *   Updated plugin data.
+   */
+  private function getDerivativeValues(array $base_plugin_definition, $url_long, $url) {
```

Comment:
80% of this function could be reduced with array and foreach.
Is not critical, but will make me happy :)

Reply:
It's 2020 - no happiness )))